File: apple/RemoteDisplay/Services/ServiceController.swift
//
//  ServiceController.swift
//  RemoteDisplay
//
//  Created by Júlio Flores on 14/09/2026.
//

extension Services {
	@ServiceActor
	final class ServiceController {
		let builtIn: [any Service]

		@Persist(key: "services.simulator.autoConnect")
		var autoConnect = false

		private(set) var status: Services.Status = .unavailable

		private var currentIndex = 0

		init(dependencies: Dependencies) throws {
			self.builtIn = dependencies.builtInServices
			guard !self.builtIn.isEmpty else {
				throw NoBuiltInServiceError()
			}
			if let autoConnect = dependencies.autoConnect {
				self.autoConnect = autoConnect
			}
			if self.autoConnect {
				dependencies.task { @ServiceActor @Sendable in
					let services = self.search()
					if let firstAvailable = await services.firstAvailable {
						try await firstAvailable.connect()
					}
				}
			}
		}

		func search() -> AsyncStream<any Service> {
			AsyncStream { continuation in
				Task {
					// it's ok to swallow the error;
					// the consumer will check individual service statuses
					try await withThrowingTaskGroup(of: Void.self) { group in
						self.status = .searching
						for service in self.builtIn {
							group.addTask {
								try await ServiceActor.run {
									try await service.search()
									continuation.yield(service)

									if service.status == .available {
										self.status = .available
									}
								}
							}
						}

						try await group.waitForAll()
						if self.status != .available {
							self.status = .unavailable
						}
						continuation.finish()
					}
				}
			}
		}

		func connect() async throws {
			self.status = .connecting
			try await self.builtIn[self.currentIndex].connect()
			self.status = .connected
		}

		@discardableResult
		func selectFirstService() throws -> any Service {
			guard self.status != .connecting && self.status != .connected else {
				throw ChangeServiceWhileConnectedError()
			}

			self.currentIndex = 0
			return self.builtIn[self.currentIndex]
		}

		@discardableResult
		func selectNextService() throws -> (any Service)? {
			guard self.status != .connecting && self.status != .connected else {
				throw ChangeServiceWhileConnectedError()
			}

			guard self.currentIndex < self.builtIn.count - 1 else {
				return nil
			}

			self.currentIndex += 1
			return self.builtIn[self.currentIndex]
		}

		func disconnect() async throws {
			self.status = .disconnecting
			try await self.builtIn[self.currentIndex].disconnect()
			self.status = .disconnected
		}
	}
}

private extension AsyncStream where Element == any Services.Service {
	var firstAvailable: Element? {
		get async {
			await first { @ServiceActor service in
				service.status == .available
			}
		}
	}
}

File: android/app/src/main/java/dk/cipher/remotedisplay/services/ServiceController.kt
package dk.cipher.remotedisplay.services

import android.content.Context
import android.content.SharedPreferences
import androidx.core.content.getSystemService
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.launch
import androidx.core.content.edit
import kotlinx.coroutines.coroutineScope

class ServiceController(dependencies: Dependencies) {
    val builtIn: List<Services.Service>
    var autoConnect: Boolean
        get() = _autoConnect
        set(value) {
            _autoConnect = value
            this.storage?.edit {
                putBoolean(AUTO_CONNECT_STORAGE_KEY, value)
            }
        }

    var status: Services.Status = Services.Status.Unavailable
        private set

    private val storage: SharedPreferences?
    private var connectedIndex = 0
    private var _autoConnect: Boolean

    init {
        this.builtIn = dependencies.builtInServices
        if (this.builtIn.isEmpty()) {
            throw Services.NoBuiltInServiceError()
        }
        this.storage = dependencies.context.getSystemService<SharedPreferences>()
        if (dependencies.autoConnect != null) {
            _autoConnect = dependencies.autoConnect
        } else {
            _autoConnect = this.storage?.getBoolean(AUTO_CONNECT_STORAGE_KEY, false) ?: false
        }

        if (this.autoConnect) {
            dependencies.task {
                val services = this.search()
                val firstServiceAvailable = services.firstAvailable()
                if (firstServiceAvailable != null) {
                    firstServiceAvailable.connect()
                }
            }
        }
    }

    fun search(): Channel<Services.Service> {
        val channel = Channel<Services.Service>()
        CoroutineScope(SupervisorJob()).launch {
            coroutineScope {
                for (service in builtIn) {
                    launch(Dispatchers.IO) {
                        service.search()
                        channel.send(service)

                        if (service.status == Services.Status.Available) {
                            status = Services.Status.Available
                        }
                    }
                }
            }
            if (status != Services.Status.Available) {
                status = Services.Status.Unavailable
            }
            channel.close()
        }
        return channel
    }

    suspend fun connect() {
        this.status = Services.Status.Connecting
        this.builtIn[this.connectedIndex].connect()
        this.status = Services.Status.Connected
    }

    fun selectFirstService(): Services.Service {
        if (this.status == Services.Status.Connecting || this.status == Services.Status.Connected) {
            throw Services.ChangeServiceWhileConnectedError()
        }
        this.connectedIndex = 0
        return this.builtIn[this.connectedIndex]
    }

    fun selectNextService(): Services.Service? {
        if (this.status == Services.Status.Connecting || this.status == Services.Status.Connected) {
            throw Services.ChangeServiceWhileConnectedError()
        }
        if (this.connectedIndex >= this.builtIn.size - 1) {
            return null
        }
        this.connectedIndex += 1
        return this.builtIn[this.connectedIndex]
    }

    suspend fun disconnect() {
        this.status = Services.Status.Disconnecting
        this.builtIn[this.connectedIndex].disconnect()
        this.status = Services.Status.Disconnected
    }

    data class Dependencies(
        val builtInServices: List<Services.Service>,
        val context: Context,
        val task: ConcurrencyContext,
        val autoConnect: Boolean?
    ) {
        typealias ConcurrencyContext = (suspend () -> Unit) -> Unit

        companion object {
            fun live(context: Context) = Dependencies(
                builtInServices = listOf(
//                    ServiceSimulator("Simulated Service LALA", context),
//                    ServiceSimulator("Simulated Service LONES", context)
                ),
                context = context,
                task = { operation ->
                    CoroutineScope(Dispatchers.Main).launch {
                        operation()
                    }
                },
                autoConnect = null
            )
        }
    }

    companion object {
        const val AUTO_CONNECT_STORAGE_KEY = "services.simulator.autoConnect"
    }
}

suspend fun Channel<Services.Service>.firstAvailable(): Services.Service? {
    val iterator = this.iterator()
    while (iterator.hasNext()) {
        val service = iterator.next()
        if (service.status == Services.Status.Available) {
            return service
        }
    }
    return null
}